### PR TITLE
Add linear gradgrad

### LINKED ISSUE #92206: RuntimeError: derivative for aten::mps_linear_backward is not implemented
### 🐛 Describe the bug

when I calculate the second  derivative of the model, using my MacBook Air with M1, the bug is shown as below
image

Another question: what amazes me most is that when I using the other computer with Windows PyTorch on cuda, the codes can run successfully...

I just want to know how to solve the problem on my MacBook shown in the picture above.

### Versions

     15 u_pred = model(f_tmp)
     16 du_df=torch.autograd.grad(outputs =u_pred, inputs = f_tmp,
     17                             grad_outputs = torch.ones_like(u_pred).to(device),
     18                             create_graph = True,
     19                             allow_unused=True
     20                             )[0]
---> 21 ddu_ddf=torch.autograd.grad(outputs =du_df, inputs = f_tmp,
     22                             grad_outputs = torch.ones_like(du_df).to(device),
     23                             create_graph = True,
     24                             allow_unused=True
     25                             )[0]

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @kulinseth @malfet @DenisVieriu97 @razarmehr @abhudev

### PR DESCRIPTION
**Pull request by qqaatw**
Fixes #92206

### CODE REVIEW

File: tools/autograd/derivatives.yaml
Comment (soulitzer) [reply tools/autograd/derivatives.yaml:2178]: To avoid adding yet another ternary, we should move this formula to FunctionsManual.cpp.

File: tools/autograd/derivatives.yaml
Comment (kulinseth) [inline tools/autograd/derivatives.yaml:2180]: I am concerned about the performance of these layers. @qqaatw , can you please benchmark these layers and check the performance ?

File: tools/autograd/derivatives.yaml
Comment (qqaatw) [reply tools/autograd/derivatives.yaml:2178]: The test failures in CI seem to be due to an undefined `grad_output` being unpacked in the newly generated `linear_backward` function in `VariableType*.cpp` (Before the gradgrad added in `derivatives.yaml`, this function is `torch::autograd::autogradNotImplementedFallback()`). I'm not familiar with nested tensors and this problem seems to only with this kind of tensors.


```c++
::std::tuple<at::Tensor,at::Tensor,at::Tensor> linear_backward(c10::DispatchKeySet ks, const at::Tensor & self, const at::Tensor & grad_output, const at::Tensor & weight, ::std::array<bool,3> output_mask) {
  auto& self_ = unpack(self, "self", 0);
  auto& grad_output_ = unpack(grad_output, "grad_output", 1);  // Here.
  auto& weight_ = unpack(weight, "weight", 2);
  auto _any_requires_grad = compute_requires_grad( self, grad_output, weight );
  ...
```

```
Exception raised from checked_cast_variable at ../pytorch/torch/csrc/autograd/VariableTypeManual.cpp:61 (most recent call first):
frame #0: decltype(static_cast<c10::(anonymous namespace)::GetFetchStackTrace()::$_0&>(fp)()) std::__1::__invoke<c10::(anonymous namespace)::GetFetchStackTrace()::$_0&>(c10::(anonymous namespace)::GetFetchStackTrace()::$_0&) + 32 (0x103db973c in libc10.dylib)
frame #1: std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > std::__1::__invoke_void_return_wrapper<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, false>::__call<c10::(anonymous namespace)::GetFetchStackTrace()::$_0&>(c10::(anonymous namespace)::GetFetchStackTrace()::$_0&) + 40 (0x103db96ec in libc10.dylib)
frame #2: std::__1::__function::__alloc_func<c10::(anonymous namespace)::GetFetchStackTrace()::$_0, std::__1::allocator<c10::(anonymous namespace)::GetFetchStackTrace()::$_0>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > ()>::operator()() + 40 (0x103db96b8 in libc10.dylib)
frame #3: std::__1::__function::__func<c10::(anonymous namespace)::GetFetchStackTrace()::$_0, std::__1::allocator<c10::(anonymous namespace)::GetFetchStackTrace()::$_0>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > ()>::operator()() + 36 (0x103db8098 in libc10.dylib)
frame #4: std::__1::__function::__value_func<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > ()>::operator()() const + 72 (0x103dc6014 in libc10.dylib)
frame #5: std::__1::function<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > ()>::operator()() const + 32 (0x103db6300 in libc10.dylib)
frame #6: c10::Error::Error(c10::SourceLocation, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >) + 76 (0x103db6860 in libc10.dylib)
frame #7: c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) + 112 (0x103db2b84 in libc10.dylib)
frame #8: torch::autograd::VariableType::(anonymous namespace)::checked_cast_variable(at::Tensor const&, char const*, int) + 152 (0x2b64605b0 in libtorch_cpu.dylib)
frame #9: torch::autograd::VariableType::unpack(at::Tensor const&, char const*, int) + 40 (0x2b646050c in libtorch_cpu.dylib)
frame #10: torch::autograd::VariableType::(anonymous namespace)::linear_backward(c10::DispatchKeySet, at::Tensor const&, at::Tensor const&, at::Tensor const&, std::__1::array<bool, 3ul>) + 128 (0x2b314d880 in libtorch_cpu.dylib)
frame #11: c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<std::__1::tuple<at::Tensor, at::Tensor, at::Tensor> (c10::DispatchKeySet, at::Tensor const&, at::Tensor const&, at::Tensor const&, std::__1::array<bool, 3ul>), &(torch::autograd::VariableType::(anonymous namespace)::linear_backward(c10::DispatchKeySet, at::Tensor const&, at::Tensor const&, at::Tensor const&, std::__1::array<bool, 3ul>))>, std::__1::tuple<at::Tensor, at::Tensor, at::Tensor>, c10::guts::typelist::typelist<c10::DispatchKeySet, at::Tensor const&, at::Tensor const&, at::Tensor const&, std::__1::array<bool, 3ul> > >, std::__1::tuple<at::Tensor, at::Tensor, at::Tensor> (c10::DispatchKeySet, at::Tensor const&, at::Tensor const&, at::Tensor const&, std::__1::array<bool, 3ul>)>::call(c10::OperatorKernel*, c10::DispatchKeySet, at::Tensor const&, at::Tensor const&, at::Tensor const&, std::__1::array<bool, 3ul>) + 408 (0x2b3150860 in libtorch_cpu.dylib)
frame #12: std::__1::tuple<at::Tensor, at::Tensor, at::Tensor> c10::callUnboxedKernelFunction<std::__1::tuple<at::Tensor, at::Tensor, at::Tensor>, at::Tensor const&, at::Tensor const&, at::Tensor const&, std::__1::array<bool, 3ul> >(void*, c10::OperatorKernel*, c10::DispatchKeySet, at::Tensor const&, at::Tensor const&, at::Tensor const&, std::__1::array<bool, 3ul>&&) + 196 (0x2b012d2b4 in libtorch_cpu.dylib)
frame #13: at::_ops::linear_backward::call(at::Tensor const&, at::Tensor const&, at::Tensor const&, std::__1::array<bool, 3ul>) + 1604 (0x2b012c038 in libtorch_cpu.dylib)
frame #14: at::linear_backward(at::Tensor const&, at::Tensor const&, at::Tensor const&, std::__1::array<bool, 3ul>) + 104 (0x2b293483c in libtorch_cpu.dylib)
frame #15: torch::autograd::generated::LinearBackward0::apply(std::__1::vector<at::Tensor, std::__1::allocator<at::Tensor> >&&) + 692 (0x2b28dea0c in libtorch_cpu.dylib)
frame #16: torch::autograd::Node::operator()(std::__1::vector<at::Tensor, std::__1::allocator<at::Tensor> >&&) + 664 (0x2b41bc840 in libtorch_cpu.dylib)
frame #17: torch::autograd::call_function(std::__1::shared_ptr<torch::autograd::GraphTask>&, torch::autograd::Node*, torch::autograd::InputBuffer&) + 680 (0x2b41b8e78 in libtorch_cpu.dylib)
frame #18: torch::autograd::Engine::evaluate_function(std::__1::shared_ptr<torch::autograd::GraphTask>&, torch::autograd::Node*, torch::autograd::InputBuffer&, std::__1::shared_ptr<torch::autograd::ReadyQueue> const&) + 924 (0x2b41a7ee0 in libtorch_cpu.dylib)
frame #19: torch::autograd::Engine::thread_main(std::__1::shared_ptr<torch::autograd::GraphTask> const&) + 1132 (0x2b4197778 in libtorch_cpu.dylib)
frame #20: torch::autograd::Engine::execute_with_graph_task(std::__1::shared_ptr<torch::autograd::GraphTask> const&, std::__1::shared_ptr<torch::autograd::Node>, torch::autograd::InputBuffer&&) + 536 (0x2b4196da4 in libtorch_cpu.dylib)
frame #21: torch::autograd::python::PythonEngine::execute_with_graph_task(std::__1::shared_ptr<torch::autograd::GraphTask> const&, std::__1::shared_ptr<torch::autograd::Node>, torch::autograd::InputBuffer&&) + 112 (0x15178d40c in libtorch_python.dylib)
frame #22: torch::autograd::Engine::execute(std::__1::vector<torch::autograd::Edge, std::__1::allocator<torch::autograd::Edge> > const&, std::__1::vector<at::Tensor, std::__1::allocator<at::Tensor> > const&, bool, bool, bool, std::__1::vector<torch::autograd::Edge, std::__1::allocator<torch::autograd::Edge> > const&) + 1948 (0x2b4196974 in libtorch_cpu.dylib)
frame #23: torch::autograd::python::PythonEngine::execute(std::__1::vector<torch::autograd::Edge, std::__1::allocator<torch::autograd::Edge> > const&, std::__1::vector<at::Tensor, std::__1::allocator<at::Tensor> > const&, bool, bool, bool, std::__1::vector<torch::autograd::Edge, std::__1::allocator<torch::autograd::Edge> > const&) + 180 (0x15178d328 in libtorch_python.dylib)
```

Comment (qqaatw) [reply tools/autograd/derivatives.yaml:2178]: Do you have any idea on how to cope with it?

File: tools/autograd/derivatives.yaml
Comment (qqaatw) [reply tools/autograd/derivatives.yaml:2180]: This is higher order gradient support where I think most model/training won't access it. If the performance impact is significant, we can maybe branch on `GradMode::is_enabled()` like the below example for fused and infinitely differentiable paths, but I need some time to dig into those so would be inclined to get this PR merged first to unblock some users.

```
- name: logit(Tensor self, float? eps=None) -> Tensor
  self: "GradMode::is_enabled() ? infinitely_differentiable_logit_backward(grad, self, eps) : logit_backward(grad, self, eps)"
  result: auto_element_wise
```

File: tools/autograd/derivatives.yaml
Comment (soulitzer) [reply tools/autograd/derivatives.yaml:2178]: @qqaatw updating the formula for linear to be the following should fix that:
```
- name: linear(Tensor input, Tensor weight, Tensor? bias=None) -> Tensor
  input, weight, bias: "grad.defined() ? linear_backward(input, grad, weight, grad_input_mask) :  std::tuple<Tensor, Tensor, Tensor>()"
  ```

File: test/test_mps.py
Comment (qqaatw) [reply test/test_mps.py]: Lowered.

File: test/test_mps.py
Comment (soulitzer) [inline test/test_mps.py:1482]: Can we also check gradgradcheck on bias?
Also to simplify things we can call to gradgradcheck with all the inputs passed in at the same time: input, weight, bias. It would also check more combinations of undefined grad handling.

Comment (qqaatw) [reply test/test_mps.py:1482]: Yes indeed, the bias term should be considered. I will update that!

File: torch/testing/_internal/common_methods_invocations.py
Comment (qqaatw) [reply torch/testing/_internal/common_methods_invocations.py:12727]: Yes, you're right.

File: test/test_mps.py
Comment (qqaatw) [inline test/test_mps.py:1488]:
```suggestion
            gradgradcheck(
                torch.nn.functional.linear,
                inputs,
                atol=gradgrad_atol,
                rtol=10.4e-05,
                raise_exception=True,
                fast_mode=False
            )
```

File: test/test_mps.py
Comment (qqaatw) [inline test/test_mps.py:1518]: The reasons these errors pile up again are likely because 1. MPS only supports float32 2. the fused `_mps_linear` computes somewhat differently from the cpu that uses composite. More context provided by another user: https://github.com/pytorch/pytorch/issues/77764#issuecomment-1469930277

Comment (albanD) [reply test/test_mps.py:1518]: These tolerances definitely look too big. We should find a better way to test this.
3e-1 is the order of magnitude of the error you would get if your formula was wrong here.

File: test/test_mps.py
Comment (albanD) [inline test/test_mps.py:1488]: I don't think you should be using gradcheck in fp32 tbh.
You might as well do the backward for a random set of gradient compared to CPU and make sure these match.
Even for a reduction op like this one, it should give you meaningfully similar results.

Another option is to check only the gradient (not double backprop) of `torch.ops.aten.linear_backward`.

Comment (qqaatw) [reply test/test_mps.py]: Thanks a lot! So would you suggest I remove the gradgradcheck at all or keep it as is?

Comment (albanD) [reply test/test_mps.py]: Yes we should remove the gradgradcheck test and all the tolerance stuff yes.

Comment (qqaatw) [reply test/test_mps.py]: Done, please take a look.

**Review by albanD — APPROVED** [review]
Change looks good. Let's make sure CI is green before merging this!

**Pull request closed without merging**
